src/journal-remote/journal-upload.c

```diff
@@ -302,10 +438,37 @@ static size_t fd_input_callback(void *buf, size_t size, size_t nmemb, void *user
 
         assert(!size_multiply_overflow(size, nmemb));
 
-        n = read(u->input, buf, size * nmemb);
-        log_debug("%s: allowed %zu, read %zd", __func__, size*nmemb, n);
-        if (n > 0)
-                return n;
+        read_size = size * nmemb;
+        if (read_size > u->bytes_left)
+                read_size = u->bytes_left;
+
+        if (u->compression != COMPRESSION_NONE) {
+                compression_buffer = malloc(read_size);
+                if (!compression_buffer) {
+                        log_oom();
+                        return CURL_READFUNC_ABORT;
+                }
+        }
+
+        n = read(u->input, compression_buffer ?: buf, read_size);
+        log_debug("%s: allowed %zu, read %zd", __func__, read_size, n);
+
+        if (n > 0) {
+                update_uploader_stat(u, n);
+                if (u->compression == COMPRESSION_NONE) {
+                        return n;
+                }
+
+                size_t compressed_size;
+
+                r = compress_blob(u->compression, compression_buffer, n, buf, read_size, &compressed_size, u->compression_level);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to compress %ld bytes using Compression=%s, CompressionLevel=%d)",
```

Comment:
always suffix log messages that are about some error code you receive with `: %m`

Reply:
updated